cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -203,6 +204,32 @@ public long getProgramRunCount(ProgramId programId) throws Exception {
       .collect(Collectors.toList());
   }
 
+  /**
+   * Get the latest run within the specified start and end times for the specified program.
+   *
+   * @param programId the program to get runs for
+   * @param programRunStatus status of runs to return
+   * @param start earliest start time of runs to return
+   * @param end latest start time of runs to return
+   * @param limit the maximum number of runs to return
+   * @return the latest runs for the program sorted by start time, with the newest run as the first run
+   * @throws NotFoundException if the application to which this program belongs was not found or the program is not
+   *                           found in the app
+   * @throws UnauthorizedException if the principal does not have access to the program
+   * @throws Exception if there was some other exception performing authorization checks
+   */
+  public List<RunRecord> getRuns(ProgramId programId, ProgramRunStatus programRunStatus,
+                                 long start, long end, int limit) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
```

Comment:
this will make a REST call for each program to make the auth check, there is an `isVisible()` method in `AuthorizationEnforcer` which takes a collection of the entity ids and do the auth check in batch. I think what we did for run count is to have the method `getProgramRunCounts` take a list of programs and doing all the operations in batch instead of starting the transaction for each operation.